### PR TITLE
Users/dikhakha/add web config azure

### PR DESCRIPTION
**Pull request by dikhakha**

### CODE REVIEW

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 1]
```diff
@@ -9,6 +10,7 @@ var utility = require('webdeployment-common/utility.js');
 var msDeploy = require('webdeployment-common/deployusingmsdeploy.js');
 var fileTransformationsUtility = require('webdeployment-common/fileTransformationsUtility.js');
 var kuduUtility = require('webdeployment-common/kuduutility.js');
+var generateWebConfigUtil = require('webdeployment-common/generatewebconfig.js');
```
Comment (sachinma) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: use camel casing in filename.

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 2]
```diff
@@ -70,9 +74,44 @@ async function run() {
...
 
         var isFolderBasedDeployment = utility.isInputPkgIsFolder(webDeployPkg);
+        var applyFileTransformFlag = JSONFiles.length != 0 || xmlTransformation || xmlVariableSubstitution;
 
-        if(JSONFiles.length != 0 || xmlTransformation || xmlVariableSubstitution) {
-            var output = await fileTransformationsUtility.fileTransformations(isFolderBasedDeployment, JSONFiles, xmlTransformation, xmlVariableSubstitution, webDeployPkg);
+        if (applyFileTransformFlag || generateWebConfig) {
+            var folderPath = await utility.generateTemporaryFolderForDeployment(isFolderBasedDeployment, webDeployPkg);
+
+            if (generateWebConfig) {
```
Comment (sachinma) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: Yon can move this logic to a private func

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 3]
```diff
@@ -70,9 +74,44 @@ async function run() {
...
+
+                    tl.debug(tl.loc('WebConfigDoesNotExist'));
+
+                    //Extract out the appType parameter as it is not to be replaced.
+                    var webConfigParameters = ParameterParser.parse(webConfigParametersStr);
+                    var appType: string = webConfigParameters['appType'].value;
+                    delete webConfigParameters['appType'];
+
+                    // Get the template path for the given appType
+                    var webConfigTemplatePath = path.join(__dirname, path.normalize('node_modules/webdeployment-common/WebConfigTemplates'), appType.toLowerCase());
```
Comment (sachinma) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: Move this logic to generateWebConfigUtil. The function should take the app name.

File: Tasks/Common/webdeployment-common/WebConfigTemplates/node [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        <!-- First we consider whether the incoming URL matches a physical file in the /public folder -->
+        <rule name="StaticContent">
+          <action type="Rewrite" url="public{REQUEST_URI}"/>
+        </rule>
+
+        <!-- All other URLs are mapped to the node.js site entry point -->
+        <rule name="DynamicContent">
+          <conditions>
+            <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="True"/>
+          </conditions>
```
Comment (sachinma) [inline Tasks/Common/webdeployment-common/WebConfigTemplates/node]: Try requesting a js file directly; and see that it is blocked or it goes via server.js

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 5]
```diff
@@ -70,9 +74,44 @@ async function run() {
...
-            var output = await fileTransformationsUtility.fileTransformations(isFolderBasedDeployment, JSONFiles, xmlTransformation, xmlVariableSubstitution, webDeployPkg);
+        if (applyFileTransformFlag || generateWebConfig) {
+            var folderPath = await utility.generateTemporaryFolderForDeployment(isFolderBasedDeployment, webDeployPkg);
+
+            if (generateWebConfig) {
+
+                //Generate the web.config file if it does not already exist.
+                var webConfigPath = path.join(folderPath, "web.config");
+                if (!fs.existsSync(webConfigPath)) {
+
```
Comment (sachinma) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: consider using tl.exist set of functions

File: Tasks/AzureRmWebAppDeployment/ParameterParser.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
+// TODO: This whole file is copied from AzureResourceGroupDeployment task. Move this parser to common lib
+// Similar parser is used in Grid UI extension. Try to move the code to some place where all can use it.
+
+export function parse(input: string) {
```
Comment (vincentdass) [inline Tasks/AzureRmWebAppDeployment/ParameterParser.ts]: Can u please add L0 test for this file separately ?

File: Tasks/Common/webdeployment-common/WebConfigTemplates/node (hunk 4 above)
Comment (sachinma) [inline Tasks/Common/webdeployment-common/WebConfigTemplates/node]: What is the observed behavior here?

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 7]
```diff
@@ -116,6 +116,28 @@
...
+            "required": false,
+            "groupName": "FileTransformsAndVariableSubstitution",
+            "helpMarkDown": "A standard Web.config will be generated and deployed to Azure App Service if the application does not have one. The values in web.config can be edited and vary based on the application framework. For example for node.js application, web.config will have startup file and iis_node module values. [Learn more](https://go.microsoft.com/fwlink/?linkid=843469)."
+        },
+        {
+            "name": "WebConfigParameters",
+            "type": "multiLine",
+            "label": "Edit Web.config parameters",
+            "required": true,
+            "defaultValue": "-appType Node -Handler iisnode -NodeStartFile server.js",
```
Comment (sachinma) [inline Tasks/AzureRmWebAppDeployment/task.json]: Lets have an empty default value.

File: Tasks/AzureRmWebAppDeployment/ParameterParser.ts (hunk 6 above)
Comment (dikhakha) [inline Tasks/AzureRmWebAppDeployment/ParameterParser.ts]: this file is taken from VSO/DistributedTask/Service/WebAccess/DistributedTask/Scripts/Extensions/Common/PowerShellParameters.ts which contains L0 tests so I have not added it here

File: Tasks/AzureRmWebAppDeployment/ParameterParser.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
+// TODO: This whole file is copied from AzureResourceGroupDeployment task. Move this parser to common lib
```
Comment (dikhakha) [inline Tasks/AzureRmWebAppDeployment/ParameterParser.ts]: will take it up in next iteration as discussed.

**Pull request merged**